ext/opentelemetry-exporter-cloud-monitoring/src/opentelemetry/exporter/cloud_monitoring/__init__.py

```diff
@@ -33,6 +46,9 @@ def __init__(self, project_id=None, client=None):
         self.project_name = self.client.project_path(self.project_id)
         self._metric_descriptors = {}
         self._last_updated = {}
+        self.unique_identifier = None
+        if add_unique_identifier:
+            self.unique_identifier = uuid.uuid4().hex[:8]
```

Comment:
I think each metric is global across all threads, so multithreading wouldn't be an issue. Multiprocessing would be an issue though -- I believe there is some discussion on adding a process identifier to the resource which would give you one time series per subprocess.

Reply:
Thanks for the link to the cardinality stuff, I agree that this approach will cause problems there. The issue is with adding an identifier on a per machine basis is that its not granular enough. I'm not sure about multithreading and whether it would result in collisions (I had thought before that it would, since I know each thread should have one (different) exporter if you have multiple meters). But if you were to run the same server concurrently on the same computer having a machine identifier wouldn't work. If you were have an identifier on the process level there would be the same cardinality issue. 

In terms of even setting a machine identifier, https://cloud.google.com/monitoring/custom-metrics/creating-metrics#which-resource are the available MonitoredResources, of which generic_node seems most suitable? Either way this will need to be done in the Resources side of things.

Reply:
The cardinality here is about the same as PID, I think the cloud monitoring docs are specifically warning against using label values that are unique per request, e.g. request IDs.

@aabmass do you expect PID cardinality to be lower because the range of PIDs is smaller and we'll end up reusing PIDs over time? This sounds especially likely to me in GKE and other containerized envs where there are few processes running per machine. If we do reuse PIDs we have another problem though: metrics from different processes at different times will get lumped together in cloud monitoring.

Reply:
@AndrewAXue I think its fine to have one exporter for all threads? You set the exporter on the `MeterProvider` which is just a global variable, not a thread local.
https://github.com/open-telemetry/opentelemetry-python/blob/99cb04687dbc606d8226a6e0a5b1ae7dde9a2174/opentelemetry-api/src/opentelemetry/metrics/__init__.py#L462

I guess for multiple processes there are a few cases:
- running containers of a single process application potentially on the same host - each container gets its own hostname I believe, so you could pass it in [`generic_node.node_id`](https://cloud.google.com/monitoring/api/resources#tag_generic_node). If it's k8s, that list of resources has `k8s_container` too. I am not very familiar with k8s so could be off here.
- using python `multiprocessing` - something like [`Process.name`](https://docs.python.org/3/library/multiprocessing.html?highlight=process#multiprocessing.Process.name) would have much lower cardinality. Not sure how we could instrument this though.
- running a bunch of processes directly in the OS - I would think this is more rare? Not sure what you can do besides PIDs :/

Any idea how OC handled this or what other language exporters are doing for the 10 second limit?

@c24t do you think `Process.name` is a safe bet for multiprocessing with low cardinality? I'm not sure how to deal with the lumped together PIDs from different times, that's a very good point.

Reply:
OC used PIDs. One of the reasons I chose to use UUID's instead of PIDs is to avoid the lumping issue.

Reply:
> @c24t do you think Process.name is a safe bet for multiprocessing with low cardinality?

It sounds like a safe bet to me, but not obviously better than using the PID or a UUID. With a low-cardinality label like PID or `Process.name` we run the risk of two different processes being lumped together in cloud monitoring. With a high-cardinality label we avoid that problem, but potentially affect cloud monitoring query performance.

It sounds like @AndrewAXue would prefer a high-cardinality label and @aabmass would prefer a low-cardinality label. Unless we get more information about how this actually affects cloud trace, I don't think either is obviously better than the other.

I think it's fine to use UUIDs (or similar) for now and plan to replace this with an appropriate resource label (assuming we have one) later.

Reply:
One more unrelated comment here: any reason not to use `hex(random.randint(0, 16**8))` instead of `uuid4` here? The "UUID" isn't exactly "U" if you truncate it.